ykrt/src/mt.rs

```diff
@@ -219,215 +217,206 @@ impl MTThread {
+                    match loc.compare_exchange_weak(
+                        ls,
+                        ls.with_count(count + 1),
+                        Ordering::Relaxed,
+                        Ordering::Relaxed,
+                    ) {
+                        Ok(_) => break,
+                        Err(new_ls) => {
+                            if !new_ls.is_counting() {
+                                // Another thread has probably started tracing this Location.
+                                break;
+                            }
+                            if new_ls.count() >= count {
+                                // Although our increment hasn't worked, at least the count hasn't
+                                // gone backwards: rather than holding this thread up, let's get
+                                // back to the interpreter.
+                                break;
+                            }
+                            ls = new_ls;
+                            if spinwait.spin() {
+                                // We don't want to park this thread, so even though we can see the
+                                // count is going backwards, go back to the interpreter.
+                                break;
+                            }
                         }
-                        // This Location is still being counted, so go around the loop again and
-                        // try to apply our increment.
                     }
                 }
-            }
-            PHASE_TRACING | PHASE_TRACING_LOCK => {
-                if lp.phase() == PHASE_TRACING_LOCK {
-                    // PHASE_TRACING_LOCK requires special handling, because another thread might
-                    // have grabbed it on this Location even though this thread is tracing this
-                    // location.
-
-                    if self.inner.tracer.is_none() {
-                        // If this thread isn't tracing anything, there's little point in spinning
-                        // as the thread tracing this location is still plausibly alive.
-                        return None;
-                    }
-
-                    // We wait to grab the lock.
-                    // FIXME: this is a terrible spinlock.
-                    while lp.phase() == PHASE_TRACING_LOCK {
-                        yield_now();
-                        lp = loc.load(Ordering::Acquire);
-                    }
-
-                    if lp.phase() != PHASE_TRACING {
-                        // While waiting for the Location to move beyond PHASE_TRACING_LOCK, we
-                        // probably raced with the thread tracing this location.
-                        return None;
-                    }
+                return None;
+            } else {
+                if self.inner.tracing.is_some() {
+                    // This thread is already tracing another Location, so either another
```

Comment:
debug_assert that the HotLocation struct agrees that we are tracing?

Reply:
That wouldn't be correct: at this point in the code we want to *start* tracing this location, but we don't want to do so if this thread is tracing another location. [One way of thinking of this: we can't start PT twice if we encounter a hot inner loop.]